### PR TITLE
Publishing OracleWebCenterPortal DockerScripts

### PR DESCRIPTION
**Pull request by anandsirasagi**
Signed-off-by: asirasag <anand.sirasagi@oracle.com>

### CODE REVIEW

File: OracleWebCenterPortal/COPYRIGHT.txt [hunk 1]
```diff
@@ -0,0 +1,20 @@
+Copyright (c) 2020, Oracle Corporation.  All rights reserved.
```
Comment (Djelibeybi) [inline OracleWebCenterPortal/COPYRIGHT.txt:1]: Please remove this file. We already have a `LICENSE.txt` in the root of the repository.

File: OracleWebCenterPortal/dockerfiles/12.2.1.4/Dockerfile [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+# -------------------------------------------------------------
+RUN $JAVA_HOME/bin/java -jar /u01/$FMW_PORTAL_JAR -silent -responseFile /u01/install.file -invPtrLoc /u01/oraInst.loc -jreLoc $JAVA_HOME -ignoreSysPrereqs -force -novalidation ORACLE_HOME=$ORACLE_HOME INSTALL_TYPE="WebCenter Portal" && \
+rm /u01/$FMW_PORTAL_JAR /u01/oraInst.loc /u01/install.file /u01/oracle/oracle_common/lib/ons.jar /u01/oracle/oracle_common/modules/oracle.jdbc/simplefan.jar
+
+#Rebuilding Base image
+FROM oracle/fmw-infrastructure:12.2.1.4.0
+#
+# Maintainer
+# ----------
+LABEL maintainer="Anand Sirasagi <anand.sirasagi@oracle.com>"
```
Comment (Djelibeybi) [inline OracleWebCenterPortal/dockerfiles/12.2.1.4/Dockerfile:73]: Please see the `CONTRIBUTING.md` file for the list of LABELs that we recommend.

File: OracleWebCenterPortal/dockerfiles/12.2.1.4/Dockerfile [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+
+#
+# Install the required packages
+# -----------------------------
+USER root
+RUN env && \
+    yum install -y \
```
Comment (Djelibeybi) [inline OracleWebCenterPortal/dockerfiles/12.2.1.4/Dockerfile:80]: All of these `yum` commands should be combined into a single command, i.e. `yum install -y sudo hostname zip unzip tar procps` and you can remove `yum clean all` as the `rm -rf /var/cache/yum` does the same thing.

File: OracleWebCenterPortal/dockerfiles/12.2.1.4/README.md [hunk 4]
```diff
@@ -0,0 +1,308 @@
...
+-w /usr/bin/docker-containerd -p wa
+-w /usr/bin/docker-runc -p w
+```
+Restart Audit Daemon using below commands
+```
+$ sudo systemctl restart auditd
+ ```
+### B. Configuring Docker daemon
+
+Add the following statements to the docker daemon json file - `/etc/docker/daemon.json`
```
Comment (Djelibeybi) [inline OracleWebCenterPortal/dockerfiles/12.2.1.4/README.md:27]: Why are these changes required? What purpose do they serve for your image, because none of these seem required to me.

File: OracleWebCenterPortal/dockerfiles/README.md [hunk 5]
```diff
@@ -0,0 +1,123 @@
...
+> 3. Click Accept to accept the license agreement.
+> 4. Use following commands to pull Oracle Fusion Middleware infrastructure base image from repository :
+```
+$ docker login container-registry.oracle.com
+$ docker pull container-registry.oracle.com/middleware/fmw-infrastructure:12.2.1.4
+$ docker tag container-registry.oracle.com/middleware/fmw-infrastructure:12.2.1.4 oracle/fmw-infrastructure:12.2.1.4.0
+```
+
+
+The Oracle Database image can be pulled from the [Docker Store](https://store.docker.com/images/oracle-database-enterprise-edition) or the [Oracle Container Registry](https://container-registry.oracle.com) or you can build your own using the Dockerfiles and scripts in [GitHub Location](https://github.com/oracle/docker-images/tree/master/OracleDatabase).
```
Comment (Djelibeybi) [inline OracleWebCenterPortal/dockerfiles/README.md:62]: Please remove all references to the Docker Store images as they are no longer maintained. You should only ever reference the images on Oracle Container Registry.

File: OracleWebCenterPortal/dockerfiles/12.2.1.4/Dockerfile [hunk 6]
```diff
@@ -0,0 +1,102 @@
...
+#
+# Maintainer
+# ----------
+LABEL maintainer="Anand Sirasagi <anand.sirasagi@oracle.com>"
+
+#
+# Install the required packages
+# -----------------------------
+USER root
+RUN env && \
```
Comment (Djelibeybi) [inline OracleWebCenterPortal/dockerfiles/12.2.1.4/Dockerfile:79]: Not sure what the purpose of the `env` command here is.

File: OracleWebCenterPortal/dockerfiles/12.2.1.4/README.md [hunk 7]
```diff
@@ -0,0 +1,308 @@
+# Oracle WebCenter Portal 12.2.1.4.0 on Docker
+
+Before running an Oracle WebCenter Portal image below security configuration are recommended to be followed.
+## Docker Security Configuration
+
+### A. Configuring auditing rules for docker
```
Comment (Djelibeybi) [inline OracleWebCenterPortal/dockerfiles/12.2.1.4/README.md:6]: What is the purpose of this?

File: OracleWebCenterPortal/dockerfiles/12.2.1.4/README.md [hunk 8]
```diff
@@ -0,0 +1,308 @@
...
+-w /usr/bin/docker-runc -p w
+```
+Restart Audit Daemon using below commands
+```
+$ sudo systemctl restart auditd
+ ```
+### B. Configuring Docker daemon
+
+Add the following statements to the docker daemon json file - `/etc/docker/daemon.json`
+Change Docker default location using data-root configuration in daemon.json.
```
Comment (Djelibeybi) [inline OracleWebCenterPortal/dockerfiles/12.2.1.4/README.md:28]: Why are you changing the default location?

File: OracleWebCenterPortal/dockerfiles/12.2.1.4/README.md [hunk 9]
```diff
@@ -0,0 +1,308 @@
...
+SEARCH_APP_USER_PASSWORD=<Search User Password>
+```
+
+### 1.2. Admin Container (WCPAdminContainer)
+#### A. Creating and Running Admin Container
+
+Run the following command to create the Admin Server container:
+
+```
+$ docker run -i -t --name $ADMIN_SERVER_CONTAINER_NAME --network=WCPortalNET -p <Any Free Port>:$ADMIN_PORT -v $DATA_VOLUME:/u01/oracle/user_projects --env-file <directory>/webcenter.env.list $WCPortalImageName /bin/bash -c "/u01/oracle/container-scripts/configureOrStartAdminServer.sh; /bin/bash"
```
Comment (Djelibeybi) [inline OracleWebCenterPortal/dockerfiles/12.2.1.4/README.md:122]: Why are you running `/bin/bash` at the end? Why doesn't your container just continue to run the process it's meant to run? You should never have an open shell running because it it's not an init and can't grab signals properly.

File: OracleWebCenterPortal/dockerfiles/12.2.1.4/README.md [hunk 10]
```diff
@@ -0,0 +1,308 @@
...
+$ docker container start -i WCPAdminContainer
+```
+
+### 1.3. Portal Container (WCPortalContainer)
+
+#### A. Creating and Running Portal Container
+Run the following command to create the Portal Managed Server container:
+
+```
+$ docker run -i -t --name WCPortalContainer --network=WCPortalNET -p <Any Free Port>:$MANAGED_SERVER_PORT -v $DATA_VOLUME:/u01/oracle/user_projects --env-file <directory>/webcenter.env.list $WCPortalImageName /bin/bash -c "/u01/oracle/container-scripts/configureOrStartWebCenterPortal.sh; /bin/bash"
```
Comment (Djelibeybi) [inline OracleWebCenterPortal/dockerfiles/12.2.1.4/README.md:156]: Same issue here as above: your container should not be running `/bin/bash`

File: OracleWebCenterPortal/dockerfiles/12.2.1.4/README.md [hunk 11]
```diff
@@ -0,0 +1,308 @@
...
+
+```
+$ sudo mkdir -p /scratch/wcpdocker/volumes/es
+$ sudo chown 1000:1000 /scratch/wcpdocker/volumes/es
+```
+
+#### B. Creating and Running Elasticsearch Container
+
+```
+$ docker run -i -t --name ESContainer --network=WCPortalNET -p 9200:9200 --volumes-from WCPortalContainer -v $ES_DATA_VOLUME:/u01/esHome/esNode/data --env-file <directory>/webcenter.env.list $WCPortalImageName /bin/bash -c "/u01/oracle/container-scripts/configureOrStartElasticsearch.sh; /bin/bash"
```
Comment (Djelibeybi) [inline OracleWebCenterPortal/dockerfiles/12.2.1.4/README.md:196]: Same problem here. Please remove the `/bin/bash`.

File: OracleWebCenterPortal/dockerfiles/12.2.1.4/README.md (hunk 9 above)
Comment (Djelibeybi) [reply OracleWebCenterPortal/dockerfiles/12.2.1.4/README.md:122]: Further, why are you specifying a command on the `run` line anyway? Why doesn't your container image have the correct `ENTRYPOINT` or `CMD` so that this isn't necessary? You can use one of your many environment variables to decide which script to run automatically.

File: OracleWebCenterPortal/dockerfiles/12.2.1.4/README.md (hunk 10 above)
Comment (Djelibeybi) [reply OracleWebCenterPortal/dockerfiles/12.2.1.4/README.md:156]: Further, why are you specifying a command on the `run` line anyway? Why doesn't your container image have the correct `ENTRYPOINT` or `CMD` so that this isn't necessary? You can use one of your many environment variables to decide which script to run automatically.

File: OracleWebCenterPortal/dockerfiles/12.2.1.4/README.md (hunk 11 above)
Comment (Djelibeybi) [reply OracleWebCenterPortal/dockerfiles/12.2.1.4/README.md:196]: Further, why are you specifying a command on the `run` line anyway? Why doesn't your container image have the correct `ENTRYPOINT` or `CMD` so that this isn't necessary? You can use one of your many environment variables to decide which script to run automatically.

File: OracleWebCenterPortal/dockerfiles/12.2.1.4/README.md [hunk 12]
```diff
@@ -0,0 +1,308 @@
...
+```
+DB_DROP_AND_CREATE=false
+
+```
+
+*Create container pointing to existing domain*
+*Run the command to create Container in given Sequence (run as non root user)*
+```
+# create Admin Container
+$ docker run -i -t --name WCPAdminContainer --network=WCPortalNET -p 7001:7001 -v /scratch/wcpdocker/volumes/wcpportal:/u01/oracle/user_projects --env-file /scratch/<userid>/docker/webcenter.env.list oracle/wcportal:12.2.1.4.0 /bin/bash -c "/u01/oracle/container-scripts/configureOrStartAdminServer.sh; /bin/bash"
```
Comment (Djelibeybi) [inline OracleWebCenterPortal/dockerfiles/12.2.1.4/README.md:243]: Remove `/bin/bash` at least, but everything after the image name should really be removed.

File: OracleWebCenterPortal/dockerfiles/12.2.1.4/README.md [hunk 13]
```diff
@@ -0,0 +1,308 @@
...
+ 
+# create Portal Container
+$ docker run -i -t --name WCPortalContainer --network=WCPortalNET -p 8888:8888 -v /scratch/wcpdocker/volumes/wcpportal:/u01/oracle/user_projects --env-file /scratch/<userid>/docker/webcenter.env.list oracle/wcportal:12.2.1.4.0 /bin/bash -c "/u01/oracle/container-scripts/configureOrStartWebCenterPortal.sh; /bin/bash"
```
Comment (Djelibeybi) [inline OracleWebCenterPortal/dockerfiles/12.2.1.4/README.md:246]: Remove `/bin/bash` at least, but everything after the image name should really be removed.

File: OracleWebCenterPortal/dockerfiles/12.2.1.4/README.md [hunk 14]
```diff
@@ -0,0 +1,308 @@
...
+
+# create Elasticsearch Container
+$ docker run -i -t --name ESContainer --network=WCPortalNET -p 9200:9200 --volumes-from WCPortalContainer -v /scratch/wcpdocker/volumes/es:/u01/esHome/esNode/data --env-file /scratch/<userid>/docker/webcenter.env.list oracle/wcportal:12.2.1.4.0 /bin/bash -c "/u01/oracle/container-scripts/configureOrStartElasticsearch.sh; /bin/bash"
```
Comment (Djelibeybi) [inline OracleWebCenterPortal/dockerfiles/12.2.1.4/README.md:249]: Remove `/bin/bash` at least, but everything after the image name should really be removed.

File: OracleWebCenterPortal/dockerfiles/12.2.1.4/README.md [hunk 15]
```diff
@@ -0,0 +1,308 @@
...
+
+```
+
+*Create containers*
+
+*Run the command to create Container in given Sequence (run as non root user)*
+
+```
+# create Admin Container
+$ docker run -i -t --name WCPAdminContainer --network=WCPortalNET -p 7001:7001 -v /scratch/wcpdocker/volumes/wcpportal:/u01/oracle/user_projects --env-file /scratch/<userid>/docker/webcenter.env.list oracle/wcportal:12.2.1.4.0 /bin/bash -c "/u01/oracle/container-scripts/configureOrStartAdminServer.sh; /bin/bash"
```
Comment (Djelibeybi) [inline OracleWebCenterPortal/dockerfiles/12.2.1.4/README.md:297]: Remove `/bin/bash` at least, but everything after the image name should really be removed.

File: OracleWebCenterPortal/dockerfiles/12.2.1.4/README.md [hunk 16]
```diff
@@ -0,0 +1,308 @@
...
+ 
+# create Portal Container
+$ docker run -i -t --name WCPortalContainer --network=WCPortalNET -p 8888:8888 -v /scratch/wcpdocker/volumes/wcpportal:/u01/oracle/user_projects --env-file /scratch/<userid>/docker/webcenter.env.list oracle/wcportal:12.2.1.4.0 /bin/bash -c "/u01/oracle/container-scripts/configureOrStartWebCenterPortal.sh; /bin/bash"
```
Comment (Djelibeybi) [inline OracleWebCenterPortal/dockerfiles/12.2.1.4/README.md:300]: Remove `/bin/bash` at least, but everything after the image name should really be removed.

File: OracleWebCenterPortal/dockerfiles/12.2.1.4/README.md [hunk 17]
```diff
@@ -0,0 +1,308 @@
...
+
+# create Elasticsearch Container
+$ docker run -i -t --name ESContainer --network=WCPortalNET -p 9200:9200 --volumes-from WCPortalContainer -v /scratch/wcpdocker/volumes/es:/u01/esHome/esNode/data --env-file /scratch/<userid>/docker/webcenter.env.list oracle/wcportal:12.2.1.4.0 /bin/bash -c "/u01/oracle/container-scripts/configureOrStartElasticsearch.sh; /bin/bash"
```
Comment (Djelibeybi) [inline OracleWebCenterPortal/dockerfiles/12.2.1.4/README.md:303]: Remove `/bin/bash` at least, but everything after the image name should really be removed.

File: OracleWebCenterPortal/dockerfiles/README.md [hunk 18]
```diff
@@ -0,0 +1,123 @@
...
+| :-------: | :---: |
+| RAM       | 16GB  |
+| Disk Space| 200GB+|
+
+## 2.2 Software Requirements
+
+|       | Version                        | Command to verify version |
+| :---: | :----------------------------: | :-----------------------: |
+| OS    | Oracle Linux 7.3 or higher     | more /etc/oracle-release  |
+| Docker| Docker version 17.03 or higher | docker version           |
```
Comment (Djelibeybi) [inline OracleWebCenterPortal/dockerfiles/README.md:24]: If you're using Oracle Linux (which is great), then you should be referring to [Oracle Container Runtime for Docker](https://docs.oracle.com/en/operating-systems/oracle-linux/docker/).

**Review by Djelibeybi — CHANGES_REQUESTED** [review]

File: OracleWebCenterPortal/dockerfiles/12.2.1.4/README.md (hunk 7 above)
Comment (anandsirasagi) [reply OracleWebCenterPortal/dockerfiles/12.2.1.4/README.md:6]: This are Security configuration which are recommended to be followed . As part of Security Bug fix we have added it

File: OracleWebCenterPortal/dockerfiles/12.2.1.4/README.md (hunk 8 above)
Comment (anandsirasagi) [reply OracleWebCenterPortal/dockerfiles/12.2.1.4/README.md:28]: To avoid running out of disk space.

Comment (Djelibeybi) [reply OracleWebCenterPortal/dockerfiles/12.2.1.4/README.md:28]: Then the correct thing to document is how much space is required so that customers can ensure they have sufficient space in `/var/lib/docker`. Changing the `data-root` to a non-standard location that generally doesn't exist on most installs is not appropriate.

**Comment by anandsirasagi** [comment]
Hi @Djelibeybi 

Please review the changes.

File: OracleWebCenterPortal/dockerfiles/12.2.1.4/README.md [hunk 19]
```diff
@@ -153,7 +100,7 @@ $ docker container start -i WCPAdminContainer
 Run the following command to create the Portal Managed Server container:
 
 ```
-$ docker run -i -t --name WCPortalContainer --network=WCPortalNET -p <Any Free Port>:$MANAGED_SERVER_PORT -v $DATA_VOLUME:/u01/oracle/user_projects --env-file <directory>/webcenter.env.list $WCPortalImageName /bin/bash -c "/u01/oracle/container-scripts/configureOrStartWebCenterPortal.sh; /bin/bash"
+$ docker run -i -t --name WCPortalContainer --network=WCPortalNET -p <Any Free Port>:$MANAGED_SERVER_PORT -v $DATA_VOLUME:/u01/oracle/user_projects --env-file <directory>/webcenter.env.list $WCPortalImageName configureOrStartWebCenterPortal.sh
```
Comment (Djelibeybi) [inline OracleWebCenterPortal/dockerfiles/12.2.1.4/README.md:103]: You're still calling `configureOrStartWebCenterPortal.sh` manually here. This may not work on an orchestration platform like Kubernetes or Mesosphere. Can you not switch to using an environment variable to determine whether the container should start an admin server or the portal? Better yet, being able to automatically determine whether an admin server is already running would be great.

File: OracleWebCenterPortal/dockerfiles/12.2.1.4/README.md [hunk 20]
```diff
@@ -193,7 +140,7 @@ $ sudo chown 1000:1000 /scratch/wcpdocker/volumes/es
 #### B. Creating and Running Elasticsearch Container
 
 ```
-$ docker run -i -t --name ESContainer --network=WCPortalNET -p 9200:9200 --volumes-from WCPortalContainer -v $ES_DATA_VOLUME:/u01/esHome/esNode/data --env-file <directory>/webcenter.env.list $WCPortalImageName /bin/bash -c "/u01/oracle/container-scripts/configureOrStartElasticsearch.sh; /bin/bash"
+$ docker run -i -t --name ESContainer --network=WCPortalNET -p 9200:9200 --volumes-from WCPortalContainer -v $ES_DATA_VOLUME:/u01/esHome/esNode/data --env-file <directory>/webcenter.env.list $WCPortalImageName configureOrStartElasticsearch.sh
```
Comment (Djelibeybi) [inline OracleWebCenterPortal/dockerfiles/12.2.1.4/README.md:143]: As above, you're now starting Elasticsearch manually.

File: OracleWebCenterPortal/dockerfiles/README.md [hunk 21]
```diff
@@ -59,7 +61,7 @@ $ docker tag container-registry.oracle.com/middleware/fmw-infrastructure:12.2.1.
 ```
 
 
-The Oracle Database image can be pulled from the [Docker Store](https://store.docker.com/images/oracle-database-enterprise-edition) or the [Oracle Container Registry](https://container-registry.oracle.com) or you can build your own using the Dockerfiles and scripts in [GitHub Location](https://github.com/oracle/docker-images/tree/master/OracleDatabase).
+The Oracle Database image can be pulled from the  [Oracle Container Registry](https://container-registry.oracle.com) or you can build your own using the Dockerfiles and scripts in [GitHub Location](https://github.com/oracle/docker-images/tree/master/OracleDatabase).
```
Comment (Djelibeybi) [inline OracleWebCenterPortal/dockerfiles/README.md:64]: Please replace "in [GitHub Location]" with "from the [Oracle Database](https://github.com/oracle/docker-images/tree/master/OracleDatabase) section of this repo."

**Review by Djelibeybi — CHANGES_REQUESTED** [review]
Question: you seem to have three different components all contained in a single image: the Admin Server, the WebCenter Portal and Elasticsearch.

Why do these not each have their own container image so that each one can be customised individually? Wouldn't having just one product per image make the overall image size smaller too?

**Comment by Djelibeybi** [comment]
> All the components are part of the same product and present in same middleware home, from maintainability and patching perspective , all of them are in the same image

Fair enough. I'm sure that's great for standard installs, but it does make the container deployment more difficult because instead of having an image per component, you have to decide at runtime which component to start. It also makes the resulting container footprint larger.

**Comment by anandsirasagi** [comment]
The only way we can achieve automatic startup without overriding  was to mention which container to start in env file while running docker images. But this approach is not preferable that's why we are going with overriding with CMD

**Comment by Djelibeybi** [comment]
Fair enough.

**Review by Djelibeybi — APPROVED** [review]
